### PR TITLE
Ballot message support

### PR DESCRIPTION
**Pull request by anonymoussprocket**
Along with local message 'forging' and more test coverage.

Note one exported method is no longer `async`: `TezosNodeWriter.forgeOperations()`

UPDATE: April 26, 2019 (ba205)

Unit test coverage for Michelson Parsing is 29/36. Below are the files that fail and the bugs that cause them:

- accounts.michelson: fails because the IF_SOME macro hasn't been implemented yes, which is slightly complicated because it performs operations an optional values, which aren't cleanly delineated by Michelson's grammar

- cps_fact.michelson: fails because parameter and storage are out of order. As it turns out, even if you fix the order in preprocessing, the code will be sent by the tezos node in the same order anyway (storage before code). Semantically, they are equivalent, so this isn't a big deal

- create_add1_lists.michelson and spawn_identities.michelson: fails because of CREATE_CONTRACT creating a nested contract without the aid of outside-grammar preprocessing

- mist_wallet_current.michelson: fails because of a bug in the lexer that doesn't distinguish keywords from strings in a proper way. There's a workaround for this in moo's (the lexer's) readme that I'll look into

- multisig.michelson and token.michelson: fails because of data being empty in the PUSH <type> <data> command, and a weird bug that causes an infinite loop with expandIf if fixed.

A detailed description of improvements to the Michelson parser can be found on top of Michelson.ne.

### CODE REVIEW

**Comment by anonymoussprocket** [comment]
Note, this merge supersedes TezosNodeWriter.ts in master

**Comment by anonymoussprocket** [comment]
TezosWalletUtil.getKeysFromMnemonicAndPassphrase has a breaking change in this branch

File: grammar/tezos/Michelson.ne [hunk 1]
```diff
@@ -0,0 +1,810 @@
...
+     'CMPEQ', 'CAAR', 'CAAAR', 'CAAAAR', 'CAAAAAR', 'CAAAAAAR', 'CAAAAAAAR', 'CDDR', 'CDDDR', 'CDDDDR', 'CDDDDDR', 'CDDDDDDR', 'CDDDDDDDR',
+     'ASSERT_CMPEQ', 'ASSERT_CMPLT', 'ISNAT', 'IFCMPGT', 'IFCMPGE', 'IFCMPLT', 'IFCMPLE', 'IF_SOME', 'CADR' ],
+    data: ['Unit', 'True', 'False', 'Left', 'Right', 'Pair', 'Some', 'None', 'instruction'],
+    constantData: ['Unit', 'True', 'False', 'None', 'instruction'],
+    singleArgData: ['Left', 'Right', 'Some'],
+    doubleArgData: ['Pair'],
+    elt: "Elt",
+    number: /-?[0-9]+/,
+    word: /[a-zA-z_0-9]+/,
+    string: /"(?:\\["\\]|[^\n"\\])*"/
```
Comment (anonymoussprocket) [inline grammar/tezos/Michelson.ne]: `word` and `string` definitions might make more sense as a single object.

File: src/chain/tezos/TezosNodeWriter.ts [hunk 2]
```diff
@@ -376,133 +364,104 @@ export namespace TezosNodeWriter {
...
-    }
-
-    /**
-     * Indicates whether a reveal operation has already been done for a given account.
-     *
-     * @param {string} network Which Tezos network to go against
-     * @param {KeyStore} keyStore Key pair along with public key hash
-     * @returns {Promise<boolean>} Result
-     */
-    export async function isManagerKeyRevealedForAccount(network: string, keyStore: KeyStore): Promise<boolean> {
```
Comment (umurb) [inline src/chain/tezos/TezosNodeWriter.ts]: This is a function used in wallet/thunks in Tezori. Deleting it causes problems.

Comment (anonymoussprocket) [reply src/chain/tezos/TezosNodeWriter.ts]: Yes, this is now in https://github.com/Cryptonomic/ConseilJS/blob/feature/governance/src/chain/tezos/TezosNodeReader.ts#L101

**Pull request merged**